lucene/queries/src/java/org/apache/lucene/queries/function/valuesource/CompareNumericFunction.java

```diff
@@ -0,0 +1,131 @@
+package org.apache.lucene.queries.function.valuesource;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import java.io.IOException;
+import java.util.Map;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.queries.function.ValueSource;
+import org.apache.lucene.queries.function.FunctionValues;
+import org.apache.lucene.search.IndexSearcher;
+
+
+/**
+ * Base class for comparison operators used within if statements
+ * To Solr's if function query a 0 is considered "false", all other values are "true"
+ */
+public abstract class CompareNumericFunction extends ValueSource {
+
+  public final ValueSource lhs;
+  public final ValueSource rhs;
+
+  public CompareNumericFunction(ValueSource lhs, ValueSource rhs) {
+    this.lhs = lhs;
+    this.rhs = rhs;
+  }
+
+  // Perform the comparison, returning true or false
+  public abstract boolean compareNumeric(double lhs, double rhs);
```

Comment:
I think I like this idea in general, it would get rid of a lot of cruft. I can pass a lot in as constructor args instead of creating lots of noisy classes.

I'm not sure I want to use LongBinaryPredicate (do you mean [LongPredicate](https://docs.oracle.com/javase/8/docs/api/java/util/function/LongPredicate.html)?. Taking a glance it seems like the value in using it would be in the support for and/or/not operators in a Java-consistent way. It would be nice, though a bit out of scope for this PR, to extend this idea to the other boolean operations so that valuesource's were compatible with this API. 

Let me push up a change that makes more anonymous instances and let me know what you think (I'll model in on how and, or, xor, etc are done)